File: internal/tui/components/sidebar.go
package components

import (
	"strings"

	"github.com/charmbracelet/lipgloss"
	"github.com/mailersend/mailersend-cli/internal/tui/theme"
	"github.com/mailersend/mailersend-cli/internal/tui/types"
)

const SidebarWidth = 20

var (
	sidebarStyle = lipgloss.NewStyle().
			Width(SidebarWidth).
			Border(lipgloss.NormalBorder(), false, true, false, false).
			BorderForeground(theme.Muted).
			Padding(1, 1)

	itemStyle = lipgloss.NewStyle().
			Padding(0, 1)

	activeItemStyle = lipgloss.NewStyle().
			Bold(true).
			Foreground(theme.Primary).
			Background(theme.BgSelected).
			Padding(0, 1)

	focusedItemStyle = lipgloss.NewStyle().
				Bold(true).
				Foreground(theme.Text).
				Background(theme.Accent).
				Padding(0, 1)
)

// Sidebar is the navigation sidebar component.
type Sidebar struct {
	views   []types.ViewInfo
	active  types.ViewType
	focused bool
	height  int
}

// NewSidebar creates a new sidebar.
func NewSidebar() Sidebar {
	return Sidebar{
		views:  types.AllViews(),
		active: types.ViewDomains,
	}
}

// SetActive sets the active view.
func (s *Sidebar) SetActive(v types.ViewType) {
	s.active = v
}

// Active returns the currently active view.
func (s Sidebar) Active() types.ViewType {
	return s.active
}

// SetFocused sets whether the sidebar is focused.
func (s *Sidebar) SetFocused(focused bool) {
	s.focused = focused
}

// Focused returns whether the sidebar is focused.
func (s Sidebar) Focused() bool {
	return s.focused
}

// SetHeight sets the sidebar height.
func (s *Sidebar) SetHeight(h int) {
	s.height = h
}

// Next moves to the next view.
func (s *Sidebar) Next() {
	idx := int(s.active)
	idx = (idx + 1) % len(s.views)
	s.active = s.views[idx].Type
}

// Prev moves to the previous view.
func (s *Sidebar) Prev() {
	idx := int(s.active)
	idx--
	if idx < 0 {
		idx = len(s.views) - 1
	}
	s.active = s.views[idx].Type
}

// SetView sets the view directly.
func (s *Sidebar) SetView(v types.ViewType) {
	if int(v) >= 0 && int(v) < len(s.views) {
		s.active = v
	}
}

// Width returns the sidebar width.
func (s Sidebar) Width() int {
	return SidebarWidth + 2 // account for border
}

// View renders the sidebar.
func (s Sidebar) View() string {
	var b strings.Builder

	for _, view := range s.views {
		style := itemStyle
		prefix := "  "

		if view.Type == s.active {
			if s.focused {
				style = focusedItemStyle
			} else {
				style = activeItemStyle
			}
			prefix = "▸ "
		}

		line := prefix + view.Icon + " " + view.Label
		// Pad to full width
		line = padRight(line, SidebarWidth-4)
		b.WriteString(style.Render(line))
		b.WriteString("\n")
	}

	content := strings.TrimRight(b.String(), "\n")
	return sidebarStyle.Height(s.height - 2).Render(content)
}

func padRight(s string, width int) string {
	if len(s) >= width {
		return s
	}
	return s + strings.Repeat(" ", width-len(s))
}

File: src/tui/components/sidebar.rs
use ratatui::layout::Rect;
use ratatui::style::{Modifier, Style};
use ratatui::text::Line;
use ratatui::widgets::{Block, Borders, Padding, Paragraph};
use ratatui::Frame;

use super::pad_right;
use crate::tui::theme;
use crate::tui::views::ViewType;

pub const WIDTH: u16 = 22;

pub struct Sidebar {
    active: ViewType,
    focused: bool,
}

impl Sidebar {
    pub fn new() -> Self {
        Self {
            active: ViewType::Domains,
            focused: false,
        }
    }

    pub fn set_active(&mut self, v: ViewType) {
        self.active = v;
    }

    pub fn active(&self) -> ViewType {
        self.active
    }

    pub fn set_focused(&mut self, focused: bool) {
        self.focused = focused;
    }

    pub fn next(&mut self) {
        let idx = (self.active.index() + 1) % ViewType::ALL.len();
        self.active = ViewType::ALL[idx];
    }

    pub fn prev(&mut self) {
        let idx = if self.active.index() == 0 {
            ViewType::ALL.len() - 1
        } else {
            self.active.index() - 1
        };
        self.active = ViewType::ALL[idx];
    }

    pub fn render(&self, frame: &mut Frame, area: Rect) {
        let block = Block::new()
            .borders(Borders::RIGHT)
            .border_style(Style::new().fg(theme::MUTED))
            .padding(Padding::new(1, 1, 1, 1));

        let mut lines: Vec<Line> = Vec::new();
        for v in ViewType::ALL {
            let is_active = v == self.active;
            let prefix = if is_active { "▸ " } else { "  " };
            let text = pad_right(&format!("{}{} {}", prefix, v.icon(), v.label()), 16);
            let style = if is_active {
                if self.focused {
                    Style::new()
                        .fg(theme::TEXT)
                        .bg(theme::ACCENT)
                        .add_modifier(Modifier::BOLD)
                } else {
                    Style::new()
                        .fg(theme::PRIMARY)
                        .bg(theme::BG_SELECTED)
                        .add_modifier(Modifier::BOLD)
                }
            } else {
                Style::new()
            };
            lines.push(Line::styled(format!(" {text} "), style));
        }

        frame.render_widget(Paragraph::new(lines).block(block), area);
    }
}
